src/backends/sysfs.rs

```diff
@@ -0,0 +1,1374 @@
+        }
+
+        pub fn read_u32(&mut self) -> Result<u32> {
+            let mut content = self.read_file()?;
+            content.retain(|c| c.is_ascii_digit());
+
+            let dword = content.trim().parse::<u32>().map_err(|_| {
+                io::Error::new(io::ErrorKind::InvalidData, "Could not parse u32 value")
+            })?;
+            Ok(dword)
+        }
+
+        pub fn read_bit(&mut self) -> Result<bool> {
+            let content = self.read_file()?;
+            let bit = content.trim().parse::<bool>().map_err(|_| {
+                io::Error::new(io::ErrorKind::InvalidData, "Could not parse bool value")
+            })?;
+            Ok(bit)
+        }
+
+        pub fn read_cable_plug_type(&mut self) -> Result<CablePropertyPlugEndType> {
+            let content = self.read_file()?;
+            let plug_type = if content.contains("type-c") {
+                CablePropertyPlugEndType::UsbTypeC
+            } else if content.contains("type-a") {
+                CablePropertyPlugEndType::UsbTypeA
+            } else if content.contains("type-b") {
+                CablePropertyPlugEndType::UsbTypeB
+            } else {
+                CablePropertyPlugEndType::OtherNotUsb
+            };
+
+            Ok(plug_type)
+        }
+
+        pub fn read_cable_type(&mut self) -> Result<CablePropertyType> {
+            let content = self.read_file()?;
+            let cable_type = if content.contains("active") {
+                CablePropertyType::Active
+            } else if content.contains("passive") {
+                CablePropertyType::Passive
+            } else {
+                return Err(Error::ParseStringError {
+                    field: "cable_type".to_string(),
+                    value: content,
+                    #[cfg(feature = "backtrace")]
+                    backtrace: std::backtrace::Backtrace::capture(),
+                });
+            };
+
+            Ok(cable_type)
+        }
+
+        pub fn read_cable_mode_support(&mut self) -> Result<bool> {
+            let content = self.read_file()?;
+            let mode_support = match content.chars().next() {
+                Some('0') => false,
+                Some(_) => true,
+                None => {
+                    return Err(Error::ParseStringError {
+                        field: "cable_mode_support".to_string(),
+                        value: content,
+                        #[cfg(feature = "backtrace")]
+                        backtrace: std::backtrace::Backtrace::capture(),
+                    });
+                }
+            };
+
+            Ok(mode_support)
+        }
+
+        pub fn read_fixed_supply_pdo(
+            &mut self,
+            path: &Path,
+            src_or_sink: GetPdosSrcOrSink,
+        ) -> Result<Pd3p2FixedSupplyPdo> {
+            match src_or_sink {
+                GetPdosSrcOrSink::Source => {
+                    self.set_path(&path.join("dual_role_power").to_string_lossy())?;
```

Comment:
I'm seeing a *lot* of `to_string_lossy()` calls throughout the code, which is a bit concerning: the paths should never have invalid unicode, but if they somehow do, instead you'll get very opaque "not implemented" errors? It might be cleaner to use something like camino which has a utf-8-only path type that can be converted to a string w/o error handling.